capsules/src/sha256.rs

```diff
@@ -0,0 +1,417 @@
+const SHA_256_OUTPUT_LEN_BYTES: usize = 32;
+const NUM_ROUND_CONSTANTS: usize = 64;
+
+const ROUND_CONSTANTS: [u32; NUM_ROUND_CONSTANTS] = [
+    0x428a2f98, 0x71374491, 0xb5c0fbcf, 0xe9b5dba5, 0x3956c25b, 0x59f111f1, 0x923f82a4, 0xab1c5ed5,
+    0xd807aa98, 0x12835b01, 0x243185be, 0x550c7dc3, 0x72be5d74, 0x80deb1fe, 0x9bdc06a7, 0xc19bf174,
+    0xe49b69c1, 0xefbe4786, 0x0fc19dc6, 0x240ca1cc, 0x2de92c6f, 0x4a7484aa, 0x5cb0a9dc, 0x76f988da,
+    0x983e5152, 0xa831c66d, 0xb00327c8, 0xbf597fc7, 0xc6e00bf3, 0xd5a79147, 0x06ca6351, 0x14292967,
+    0x27b70a85, 0x2e1b2138, 0x4d2c6dfc, 0x53380d13, 0x650a7354, 0x766a0abb, 0x81c2c92e, 0x92722c85,
+    0xa2bfe8a1, 0xa81a664b, 0xc24b8b70, 0xc76c51a3, 0xd192e819, 0xd6990624, 0xf40e3585, 0x106aa070,
+    0x19a4c116, 0x1e376c08, 0x2748774c, 0x34b0bcb5, 0x391c0cb3, 0x4ed8aa4a, 0x5b9cca4f, 0x682e6ff3,
+    0x748f82ee, 0x78a5636f, 0x84c87814, 0x8cc70208, 0x90befffa, 0xa4506ceb, 0xbef9a3f7, 0xc67178f2,
+];
+
+pub struct Sha256Software<'a> {
+    state: Cell<State>,
+
+    client: OptionalCell<&'a dyn Client<'a, SHA_256_OUTPUT_LEN_BYTES>>,
+    input_data: OptionalCell<LeasableBuffer<'static, u8>>,
+    data_buffer: MapCell<[u8; SHA_BLOCK_LEN_BYTES]>,
+    buffered_length: Cell<usize>,
+    total_length: Cell<usize>,
+
+    // Used to store the hash or the hash to compare against with verify
+    output_data: Cell<Option<&'static mut [u8; SHA_256_OUTPUT_LEN_BYTES]>>,
+
+    hash_values: Cell<[u32; 8]>,
+    deferred_caller: &'a DynamicDeferredCall,
+    handle: OptionalCell<DeferredCallHandle>,
+}
+
+impl<'a> Sha256Software<'a> {
+    pub fn new(call: &'a DynamicDeferredCall) -> Sha256Software<'a> {
+        Sha256Software {
+            state: Cell::new(State::Idle),
+            client: OptionalCell::empty(),
+            input_data: OptionalCell::empty(),
+            data_buffer: MapCell::new([0; SHA_BLOCK_LEN_BYTES]),
+            buffered_length: Cell::new(0),
+            total_length: Cell::new(0),
+
+            output_data: Cell::new(None),
+            hash_values: Cell::new([0; 8]),
+
+            deferred_caller: call,
+            handle: OptionalCell::empty(),
+        }
+    }
+
+    pub fn initialize_callback_handle(&self, handle: DeferredCallHandle) {
+        self.handle.replace(handle);
+    }
+
+    pub fn busy(&self) -> bool {
+        match self.state.get() {
+            State::Idle => false,
+            _ => true,
+        }
+    }
+
+    pub fn initialize(&self) -> Result<(), ErrorCode> {
+        if !self.busy() {
+            self.buffered_length.set(0);
+            self.data_buffer.map(|b| {
+                for i in 0..SHA_BLOCK_LEN_BYTES {
+                    b[i] = 0;
+                }
+            });
+            self.hash_values.set([
+                0x6a09e667, 0xbb67ae85, 0x3c6ef372, 0xa54ff53a, 0x510e527f, 0x9b05688c, 0x1f83d9ab,
+                0x5be0cd19,
+            ]);
+            Ok(())
+        } else {
+            Err(ErrorCode::BUSY)
+        }
+    }
```

Comment:
We have been trying to remove [`init()` calls ](https://github.com/tock/tock/issues/1112), why not just call this in new()?

Reply:
I put this in a separate method because this functionality is also needed by `clear_data`. But I agree, this should not be a `pub fn`, instead `new` should invoke it.